certification/internal/k8s/deployable_by_olm.go

```diff
@@ -0,0 +1,209 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	operatorv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/errors"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/cli"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	yaml "gopkg.in/yaml.v2"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/rest"
+
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	catalogSourceNs = "openshift-marketplace"
+)
+
+var (
+	k8sconfig                     *rest.Config
+	err                           error
+	subscriptionName, packageName string
+	channel, ooNamespace          string
+	operatorGroup, catalogSource  string
+	catalogImage                  string
+	targetNamespaces              []string = []string{ooNamespace}
+)
+
+type DeployableByOlmCheck struct{}
+
+func (p *DeployableByOlmCheck) Validate(image string) (bool, error) {
+
+	// create k8s custom resources for the operator deployment
+	err = p.setUp()
+	defer p.cleanUp()
+	if err != nil {
+		return false, err
+	}
+
+	var csv *operatorv1alpha1.ClusterServiceVersion
+	var installedCSV string
+	var subs *operatorv1alpha1.Subscription
+
+	// query API server for the installed CSV field of the created subscription
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	for {
+		subs, _ = openshiftEngine.GetSubscription(subscriptionName, cli.OpenShiftCliOptions{Namespace: ooNamespace}, k8sconfig)
+		installedCSV = subs.Status.InstalledCSV
+		// if the installedCSV field is present, stop the querying
+		if len(installedCSV) > 0 {
+			log.Debug(fmt.Sprintf("Subscription.status.installedCSV is %s", installedCSV))
+			break
+		}
+		// if the context deadline exceeds, fail the check
+		if ctx.Err() != nil {
+			log.Error("failed to fetch .status.installedCSV from Subscription: ", ctx.Err())
+			return false, errors.ErrK8sApiCallFailed
+		}
+		log.Debug("InstalledCSV is not ready yet, retrying...")
+		time.Sleep(2 * time.Second)
+	}
+
+	log.Trace(fmt.Sprintf("Looking for csv %s in namespace %s", installedCSV, ooNamespace))
+
+	// query API server for CSV by name
+	ctx, cancel = context.WithTimeout(context.Background(), 90*time.Second)
+	defer cancel()
+
+	for {
```

Comment:
Same with this block of code. A separate function here would clean things up a bit.